libraries/abstractions/secure_sockets/lwip/iot_secure_sockets.c

```diff
@@ -114,6 +117,69 @@ static int8_t sockets_allocated = socketsconfigDEFAULT_MAX_NUM_SECURE_SOCKETS;
 
 /*-----------------------------------------------------------*/
 
+/*
+ * @brief Sockets close
+ */
+static void prvSocketsClose( ss_ctx_t * ctx )
+{
+    uint32_t ulProtocol;
+
+    sockets_allocated++;
+
+    /* Clean-up application protocol array. */
+    if( NULL != ctx->ppcAlpnProtocols )
+    {
+        for( ulProtocol = 0;
+             ulProtocol < ctx->ulAlpnProtocolsCount;
+             ulProtocol++ )
+        {
+            if( NULL != ctx->ppcAlpnProtocols[ ulProtocol ] )
+            {
+                vPortFree( ctx->ppcAlpnProtocols[ ulProtocol ] );
+            }
+        }
+
+        vPortFree( ctx->ppcAlpnProtocols );
+    }
+
+    if( true == ctx->enforce_tls )
+    {
+        TLS_Cleanup( ctx->tls_ctx );
+    }
+
+    if( ctx->server_cert )
+    {
+        vPortFree( ctx->server_cert );
+    }
+
+    if( ctx->destination )
+    {
+        vPortFree( ctx->destination );
+    }
+
+    vPortFree( ctx );
+}
+
+/*
+ * @brief Decrement ctx refcount and call release function if the count is 1 (
+ *        last user of the ctx)
+ */
+static void prvDecrementRefCount( ss_ctx_t * ctx )
+{
+    if( Atomic_Decrement_u32( &ctx->ulRefcount ) == 1 )
```

Comment:
If the count is at 1, we would want to close the socket, as the last remaining reference has called prvDecrementRefCount. Should this not be == 0? As Atomic_Decrement_u32( x ) , would return ( x - 1 )?

Reply:
it would return x, then decrement x, thats why i am checking for 1
static portFORCE_INLINE uint32_t Atomic_Decrement_u32( uint32_t volatile * pAddend )
{
    uint32_t ulCurrent;

    ATOMIC_ENTER_CRITICAL();

    ulCurrent = *pAddend;

    *pAddend -= 1;

    ATOMIC_EXIT_CRITICAL();

    return ulCurrent;
}

Reply:
Ah i understand. Thanks for the explanation!